File: src/main/java/org/codetome/hexameter/core/api/HexagonalGridCalculator.java
package org.codetome.hexameter.core.api;

import org.codetome.hexameter.core.api.contract.SatelliteData;
import org.codetome.hexameter.core.backport.Optional;

import java.util.List;
import java.util.Set;

/**
 * Supports advanced operations on a {@link HexagonalGrid}.
 * Operations supported:
 * <ul>
 * <li>Calculating distance between 2 {@link Hexagon}s.</li>
 * <li>Calculating movement range from a {@link Hexagon} using an arbitrary distance.</li>
 * </ul>
 * <em>Not implemented yet, but are on the roadmap:</em>
 * <ul>
 * <li>Calculating movement range with obstacles</li>
 * <li>Calculating field of view</li>
 * <li>Path finding between two {@link Hexagon}s (using obstacles)</li>
 * </ul>
 */
public interface HexagonalGridCalculator<T extends SatelliteData> {

    /**
     * Calculates the distance (in hexagons) between two {@link Hexagon} objects on the grid.
     *
     * @param hex0 hex 0
     * @param hex1 hex 1
     *
     * @return distance
     */
    int calculateDistanceBetween(Hexagon<T> hex0, Hexagon<T> hex1);

    /**
     * Returns all {@link Hexagon}s which are within <code>distance</code> (inclusive) from the {@link Hexagon}.
     *
     * @param hexagon {@link Hexagon}
     * @param distance distance
     *
     * @return {@link Hexagon}s within distance (inclusive)
     */
    Set<Hexagon<T>> calculateMovementRangeFrom(Hexagon<T> hexagon, int distance);

    /**
     * Returns the Hexagon on the grid which is at the point resulted by rotating the <code>targetHex</code>'s
     * coordinates around the <code>originalHex</code> by <code>rotationDirection</code> degrees.
     *
     * @param originalHex center hex
     * @param targetHex hex to rotate
     * @param rotationDirection direction of the rotation
     *
     * @return result
     */
    Optional<Hexagon<T>> rotateHexagon(Hexagon<T> originalHex, Hexagon<T> targetHex, RotationDirection rotationDirection);


    /**
     * Returns the {@link Set} of {@link Hexagon}s which are <code>radius</code> distance
     * from <code>centerHexagon</code>.
     *
     * @param centerHexagon center
     * @param radius radius
     *
     * @return Set of hexagons or empty set if not applicable
     */
    Set<Hexagon<T>> calculateRingFrom(Hexagon<T> centerHexagon, int radius);

    /**
     * Returns a {@link List} of {@link Hexagon}s which must be traversed in the
     * given order to go from the <code>from</code> Hexagon to the <code>to</code> Hexagon.
     *
     * @param from starting hexagon
     * @param to target hexagon
     *
     * @return List of hexagons containing the line
     */
    List<Hexagon<T>> drawLine(Hexagon<T> from, Hexagon<T> to);

    /**
     * Returns true if the <code>from</code> {@link Hexagon} is visible from the <code>to</code> Hexagon.
     *
     * @param from the Hexagon that we are testing the visibility from
     * @param to the Hexagon from which we are testing the visibility to
     *
     * @return true if hexagon is visible, false otherwise
     */
    boolean isVisible(Hexagon<T> from, Hexagon<T> to);
}

File: mixite.core/core/src/main/kotlin/org/hexworks/mixite/core/api/HexagonalGridCalculator.kt
package org.hexworks.mixite.core.api

import org.hexworks.mixite.core.api.contract.SatelliteData
import org.hexworks.mixite.core.vendor.Maybe

/**
 * Supports advanced operations on a [HexagonalGrid].
 * Operations supported:
 *
 *  * Calculating distance between 2 [Hexagon]s.
 *  * Calculating movement range from a [Hexagon] using an arbitrary distance.
 *
 * *Not implemented yet, but are on the roadmap:*
 *
 *  * Calculating movement range with obstacles
 *  * Calculating field of view
 *  * Path finding between two [Hexagon]s (using obstacles)
 *
 */
interface HexagonalGridCalculator<T : SatelliteData> {

    /**
     * Calculates the distance (in hexagons) between two [Hexagon] objects on the grid.
     *
     * @param hex0 hex 0
     * @param hex1 hex 1
     *
     * @return distance
     */
    fun calculateDistanceBetween(hex0: Hexagon<T>, hex1: Hexagon<T>): Int

    /**
     * Returns all [Hexagon]s which are within `distance` (inclusive) from the [Hexagon].
     *
     * @param hexagon [Hexagon]
     * @param distance distance
     *
     * @return [Hexagon]s within distance (inclusive)
     */
    fun calculateMovementRangeFrom(hexagon: Hexagon<T>, distance: Int): Set<Hexagon<T>>

    /**
     * Returns the Hexagon on the grid which is at the point resulted by rotating the `targetHex`'s
     * coordinates around the `originalHex` by `rotationDirection` degrees.
     *
     * @param originalHex center hex
     * @param targetHex hex to rotate
     * @param rotationDirection direction of the rotation
     *
     * @return result
     */
    fun rotateHexagon(originalHex: Hexagon<T>, targetHex: Hexagon<T>, rotationDirection: RotationDirection): Maybe<Hexagon<T>>


    /**
     * Returns the [Set] of [Hexagon]s which are `radius` distance
     * from `centerHexagon`.
     *
     * @param centerHexagon center
     * @param radius radius
     *
     * @return Set of hexagons or empty set if not applicable
     */
    fun calculateRingFrom(centerHexagon: Hexagon<T>, radius: Int): Set<Hexagon<T>>

    /**
     * Returns a [List] of [Hexagon]s which must be traversed in the
     * given order to go from the `from` Hexagon to the `to` Hexagon.
     *
     * @param from starting hexagon
     * @param to target hexagon
     *
     * @return List of hexagons containing the line
     */
    fun drawLine(from: Hexagon<T>, to: Hexagon<T>): List<Hexagon<T>>

    /**
     * Returns true if the `from` [Hexagon] is visible from the `to` Hexagon.
     *
     * @param from the Hexagon that we are testing the visibility from
     * @param to the Hexagon from which we are testing the visibility to
     *
     * @return true if hexagon is visible, false otherwise
     */
    fun isVisible(from: Hexagon<T>, to: Hexagon<T>): Boolean
}
